### DISCUSSION TITLE
Global variables

### DISCUSSION THREAD

**Discussion by denny7** [vercel/next.js#15054] [Help]
Hello! I would like to create a variable in server.js and use it everywhere in application. I need to make a call to an API and get a list (with categories) which I will use in my application. The list will not be changed so I need to make a call to the API only once and store it somewhere. I would like to make something like getStaticProps does but for the whole app (not only for one page). As I understood I can't use getStaticProps in _app.js. How can I achive that? Thanks.

**Comment by jamesmosier** [comment]
There is currently no mechanism to statically fetch site wide data from an API in `_app.js`.

You could use [`getInitialProps` inside `_app`](https://nextjs.org/docs/advanced-features/custom-app) but this would then force all of your pages to be server side rendered and opt out of static optimizations.

**Reply by denny7** [reply]
`getInitalProps` will be called for every page(and API call too). I would like to make a call to API only once at build time.

**Reply by jamesmosier** [reply]
> There is currently no mechanism to statically fetch site wide data from an API in _app.js.

**Comment by SergeyRe** [comment]
It loooks like my case. Generally i pull product data from contentful. But there is part of data that are fast changed (price,availability)  and i get it from external API. Right now i first push them to contentful and then pull again for nextjs.
the problem is that process of writing to contentful is not quick enough
The idea i dont know how to realize is to  get that data just before static generation and export as global variable and then use in process of generation.
Any idea?

**Comment by aralroca** [comment]
Maybe you can use runtime configuration for that: https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration

**Reply by denny7** [reply]
How can I make an API call and store its response in runtime configuration?

**Reply by acarolinafbarros** [reply]
@denny7 
I have the same question.
I need to store a value that is retrieved by API call in runtime. Ideas?

**Reply by Raygud** [reply]
You can save it as session storage or browser storage depending on how long you need the availability

**Comment by megetron** [comment]
having the same issue here.

one solution might be:

1. generating from the database/cms a json file and use it in your _app.js as shown [here](https://frontaid.io/blog/using-nextjs-with-json-file/)
2. optional is to use react context as shown [here](https://reacttricks.com/sharing-global-data-in-next-with-custom-app-and-usecontext-hook/)
3. build your data on build time. changing the webpack like shown [here](https://dev.to/nalanj/adding-scripts-to-next-js-n7i)
OR
```
{
  webpack: (config, { isServer }) => {
  if (isServer) {
    require('./scripts/generate-json');
  }

  return config;
}
```

**Comment by connor4312** [comment]
We can use [Symbol.for](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/for) and the Node `global` for this. (You could also use plain strings, but symbols reduce the chance of influencing things you don't mean to.) For TypeScript, I made this helper class:

```ts
export class GlobalRef<T> {
  private readonly sym: symbol;

  constructor(uniqueName: string) {
    this.sym = Symbol.for(uniqueName);
  }

  get value() {
    return (global as any)[this.sym] as T | undefined;
  }

  set value(value: T) {
    (global as any)[this.sym] = value;
  }
}
```

Then I can use it like:

```ts
const databaseConn = new GlobalRef('myapp.database');
if (!databaseConn.value) {
  databaseConn.value = setupDatabaseConnection();
}

export const database = databaseConn.value;
```

You can use this for plain JS too, just remove the type annotations :)

**Reply by jajvisser** [reply]
Thank you this solved our exact problem. 

We noticed when we called a constant from a `getServerSideProps` or `pages/api/refresh.ts` it would create 2 instances of a service. This pattern makes sure we only have 1 `service` shared between a page and api.

**Reply by glassdimlygr** [reply]
I have tested and the following works. Excuse me but... WTF 💥 . Why does it work? Why don't normal globals work? Why does this weird casting work? I usually understand things, but this I do not. What do I need to understand why this forces nextjs to share a global between routes? And furthermore... WHY DOESN'T NEXTJS SHARE A GLOBAL BETWEEN ROUTES?! WHAT MADNESS IS THIS?

> We can use [Symbol.for](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/for) and the Node `global` for this. (You could also use plain strings, but symbols reduce the chance of influencing things you don't mean to.) For TypeScript, I made this helper class:
> 
> ```ts
> export class GlobalRef<T> {
>   private readonly sym: symbol;
> 
>   constructor(uniqueName: string) {
>     this.sym = Symbol.for(uniqueName);
>   }
> 
>   get value() {
>     return (global as any)[this.sym] as T | undefined;
>   }
> 
>   set value(value: T) {
>     (global as any)[this.sym] = value;
>   }
> }
> ```
> 
> Then I can use it like:
> 
> ```ts
> const databaseConn = new GlobalRef('myapp.database');
> if (!databaseConn.value) {
>   databaseConn.value = setupDatabaseConnection();
> }
> 
> export const database = databaseConn.value;
> ```
> 
> You can use this for plain JS too, just remove the type annotations :)

**Reply by codenomnom** [reply]
This does not seem to work as of today 👀

**Reply by oldtimehacker** [reply]
I'm having trouble with globals as well, with the simplest test of simply initializing a global flag (initially set to false), then running a function that sets it to true, then later checking its status. It's false.

I'd love someone from next.js to comment on this to tell me how this can happen, why it happens, if it's by design, why, and if not, how the heck its not fixed yet?  Plz plz plz

**Comment by jacobgraf** [comment]
I need to accomplish this exact same thing. There has to be a way, right?

Like @denny7 mentioned, at build time only, I need to hit a couple of APIs, mutate the returned object a bit, cache that mutated object, and then access that object from anywhere in my app. The data (on the server/api side) will not change frequently. If/when it does, I will trigger another build cycle.

It's essentially the same functionality as getStaticProps, but for a globally-accessible variable/object.

Anyone else have any other creative solutions? Seems like this would be a pretty handy thing to have in cases where your data won't change to much and there is no need to continually re-fetch that sitewide data.

**Reply by donovanperalta** [reply]
I had to do something similar with a database connection in my Next app, since a new instance was being generated on every request. There are ways to avoid this problem using API routes with middleware, but I wanted to keep it as simple as possible. So, in your case, the solution would look something like this:

```javascript
/* next.config.js */

// The data needs to be fetched synchronously:
const fetch = require('sync-fetch');

// Get the data you need - This could be moved into a separate file
function getStaticData() {
  return fetch('https://jsonplaceholder.typicode.com/todos/1').json();
}

const staticData = getStaticData();

module.exports = () => {
  return {
    serverRuntimeConfig: { staticData }
  };
};

```

Then you can simply create an API route that will serve this static data:

```javascript
/* pages/api/static-data */

import getConfig from 'next/config';

const { serverRuntimeConfig } = getConfig();
const { staticData } = serverRuntimeConfig;

function handler(req, res) {
  res.json(staticData);
};

export default handler;
```

**Reply by LOVELYZOMBIEYHO** [reply]
It's work! but is it only fetch on the first time of loading if I am doing a dynamic navbar through connected to the backend.  I don't want to fetch every time refresh pages.

**Comment by lawrencejob** [comment]
FIrstly, Vercel and Next.js has changed front-end engineering for me, so in case an author sees this, thank you very much!

Secondly, apologies for resurrecting an old thread, but I found this while researching a solution to a problem I keep running into.

I have an `_app` with a `<Navbar>` component, which has drop-downs which populate from a call to my CMS. I don't want to put the `Navbar` into one of the pages (and optimise out the repeated calls to render that component) because I'd have to restructure the app quite significantly, and there's a lot of interesting stuff in the header that I wouldn't want to deal with the intricacies of being added and removed on page transitions.

So far, I'm using a client-side call to my CMS to get the navbar contents on page load (passing the inconvenience to the user), but this doesn't feel ideal.

There are a bunch of places both at work and on this project where it would be nice to load in (sometimes rich-) content for reuse at some global level, or the app, but keep the benefits of static rendering. It would be ideal if it turns out I'm missing something obvious, or if this is something in consideration for the project roadmap?

I've attached a mockup if it helps to visualise chat I'm stuck on:
![image](https://user-images.githubusercontent.com/1598811/149211395-7f6d676f-d83b-47e3-83d7-116118fd1905.png)

**Comment by MBelniak** [comment]
I tried using inversify and following [this example](https://stackblitz.com/edit/nextjs-with-inversify?file=src%2Fioc.ts) I managed to achieve a global database instance (at least it seems to work, but it's not field tested): 
```javascript
import { Container, injectable } from 'inversify';
import sdk from 'node-appwrite';
import { client } from '@/database/appwrite';
import 'reflect-metadata';

export abstract class AbstractDatabaseClient {
    abstract getClient(): sdk.Databases;
}
@injectable()
export class DatabaseClient implements AbstractDatabaseClient {
    private databases: sdk.Databases;
    constructor() {
        client
            .setEndpoint('https://cloud.appwrite.io/v1')
            // ... further initialization

        this.databases = new sdk.Databases(client);
    }

    getClient(): sdk.Databases {
        return this.databases;
    }
}

const container = new Container();

if (typeof window === 'undefined') {
    container
        .bind<AbstractDatabaseClient>(AbstractDatabaseClient)
        .to(DatabaseClient);
}

// Usage:
container.get(AbstractDatabaseClient).getClient();
```

**Comment by ogbotemi-2000** [comment]
Hello, I recently migrated my web app to Vercel from Railway and developed a working solution, which can be found here: *https://rmrf-css.vercel.app*, for the topic of this discussion.
Here are my observations and solutions

1. On Railway, everything works as expected of a Node.js server and the global object in my `server.js` file stores data throughout the runtime of the server.
2. Ensuring this behavior on _serverless_ Vercel required mimicking storing and reading values in objects by using the file system as shown below and reading/writing to the `/tmp/` directory to avoid errors due to a violation of writing files on the server.
3. To avoid the issue of having to serialize URLs into valid filenames for saving their responses, a uuid is generated and used as the filename of the URLs' responses, it is then sent to the client among the response for the initial request and sent back by the client to be used as a unique string bearing the name of the file to be read similar to using an object in uniqueness and versatility.

```js
/* store.js */
let fs     = require('fs'),
    uuid   = (a) => (a ? (a ^ ((Math.random() * 16) >> (a / 4))).toString(16) : ([1e7] + -1e3 + -4e3 + -8e3 + -1e11).replace(/[018]/g, uuid)),
    /* may be made asynchronous */
    rm     = id=>fs.unlinkSync(path.join(dir, id+'.txt')),
    dir    = '/tmp/',
    path   = require('path');

/*writes/reads data as text/plain, open to modifications*/
module.exports = {
  read: id =>new Promise((resolve, reject)=>fs.readFile(dir+id+'.txt', (err, buffer)=>{
      if(err) reject(err);
      else resolve(buffer.toString())
    })
  ),
  write: function(data, cb, id) {
    fs.writeFile(dir+(id=uuid())+'.txt', data, _=>cb(id))
  },
  rm 
}
```

Here is how I use it and how it may be used in the wild

```js
let store  = require('./store');
module.exports = async function(request, response, uuid) {
  let {url, id} = request.query;
  if(uuid = await store.read(id)) {
    /* exists */
    response.send(/* data */),
    /* you may remove the file or leave it since the contents of the /tmp/ directory are likely temporary */
    store.rm(uuid)
  } else {
    store.write('data', function(id) {
       /* send the unique id for the stored data in some way as JSON or text */
       response.send(id)
    })
  }
}
```

I hope it serves as a viable workaround to the issue with using global variables in Vercel serverless backends

**Comment by Plumbiu** [comment]
[instrumentation api](https://nextjs.org/docs/pages/building-your-application/optimizing/instrumentation) may be useful:

```js
export async function register() {
  if (process.env.NEXT_RUNTIME === 'nodejs') {
    // Init database or else
  }
}
```
